Norelease+flush

craff opened the pull request:
Two commits: exception in flush + functions that do not release the global loc in a separate module.

craff commented:
There is a compilation problem in a test related to Thread.exit () versus raise Thread.Exit.

we have to choose between keeping a deprecated warning in Tests (which gives an error with warn as error) or abandoning older ocaml version.  

Tell me what you prefer...

craff commented:
I fixed the problem in Test/util.ml with local [@ warning -3]

craff commented:
Here are some tests with nginx+ssl (as reference) and simple_http+ssl on a 50MB file

|number of connection |   nginx-ssl | simple-ssl  without PR |simple-ssl with PR |
|------|----------------|-------------------------------|---------------------------|
|5 |        50.30  |          36.56  |                             59.33 |
|10  | 54.95  | 32.43  | 60.96 |
|15 | 54.63  |33.81 | 61.60 |
|20 |53.06 | 33.08 | 61.57 |

This is almast a factor 2 and allow to be better that nginx!

craff commented:
No news ???

anmonteiro commented:
> No news ???

This project I maintained by (unpaid) volunteers. Please allow some time before we can review your changes (for free! Isn’t that beautiful?). 

Given that, messages like this are a bit discouraging. Rest assured your changes will be reviewed in due time.

craff commented:
Sorry for the mail showing impatience. Could you just allow the workflow build for all my PRs (or all PR?), I could check for problems myself. 

Cheers

Le 13 avril 2023 12:51:07 GMT-10:00, Antonio Nuno Monteiro ***@***.***> a écrit :
>> No news ???
>
>This project I maintained by (unpaid) volunteers. Please allow some time before we can review your changes (for free! Isn’t that beautiful?). 
>
>Given that, messages like this are a bit discouraging. Rest assured your changes will be reviewed in due time.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/savonet/ocaml-ssl/pull/106#issuecomment-1507700688
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

craff commented:
The macos build seems to loop between the ssl_context (which ends successfully) and the ssl_cipher test that never starts.

craff commented:
Apart from tests seems ok.

craff commented:
I redid a build from my github, same thing, a loop on macos. Probably not related to this patch? Not sure.

toots commented on src/ssl.mli:
```diff
@@ -471,51 +445,96 @@ val set_ip : socket -> string -> unit
+
+(** The main communication functions below comes with
+    - a "Acquire" version that acquire the runtime while the operation takes place.
+    - a "NoAcqure" version that acquire the runtime while the operation takes place.
```
```suggestion
    - a "NoAcquire" version that does not acquire the runtime while the operation takes place.
```

toots commented on src/ssl.mli:
```diff
@@ -471,51 +445,96 @@ val set_ip : socket -> string -> unit
+
+    The No-acquire version is recommanded otherwise:
+    - i.e. the program is purely sequential (one thread)
+    - the sockect is non blocking
```
```suggestion
    - the socket is non blocking
```

anmonteiro commented:
Thanks, @toots. This is still on my list and I'll try to get to it over the weekend.

toots commented:
> > Just left some comments. I don't want to overstep since we delegated this module but I would personally really like to see this work merged.
> > I've been hoping that OCaml could have a backend server with speed comparable to nginx/node ever since the work on multi-thread started and since seems to finally be bridging that gap!
> 
> Thanks for you review, I have commited the change. I guess you had a look at [simple_httpd](https://github.com/craff/)?

Not yet but I plan to.

anmonteiro commented:
> Here are some tests with nginx+ssl (as reference) and simple_http+ssl on a 50MB file
> 
> number of connection	nginx-ssl	simple-ssl without PR	simple-ssl with PR
> 5	50.30	36.56	59.33
> 10	54.95	32.43	60.96
> 15	54.63	33.81	61.60
> 20	53.06	33.08	61.57
> This PR brings almost a factor 2 and allows to be better that nginx!

What are the units here?

anmonteiro commented on src/ssl.ml:
```diff
@@ -261,91 +261,161 @@ external set_host : socket -> string -> unit = "ocaml_ssl_set1_host"
-  let domain = Unix.domain_of_sockaddr sockaddr in
-  let sock =
-    Unix.socket domain Unix.SOCK_STREAM 0 in
+module SslReleaseExt = struct
```
what does `SslReleaseExt`, `SslComExt` etc mean?

I can probably guess `SslReleaseExt` means "not releasing the runtime lock", but what's `SslComExt`?

Could you add more descriptive names, even if they're longer?

craff replied:
I addressed this by commit [cb3cc1f](https://github.com/savonet/ocaml-ssl/pull/106/commits/cb3cc1f8497fbafef4951a33b647e56b03f54c7b)

anmonteiro commented on src/ssl.mli:
```diff
@@ -471,51 +445,96 @@ val set_ip : socket -> string -> unit
+    capture all exception that requires you to retry later the same operation.
+*)
+
+module type SslCom = sig
```
we don't need to export the module type in the signature. Perhaps it'd be better to put it in a `ssl_intf.ml` and refer to it here.

craff replied:
I like to have documlentation in the same .mli file as a .ml. This is not a big mater if it is duplicated in both ssl.ml and ssl.mli ?

anmonteiro replied:
We're exposing a module type that has no use. I'd rather keep the API surface area as small as possible.

craff replied:
The module is used, twice:
```
module SslRelease : SslCom
module SslNoRelease : SslCom
```

And is the only place that gives the type of the main ssl function like `connect`, `read`, `ẁrite`, .... They are available directly as before via the include of SslRelease, I do not see how to remove `SslCom`. Or maybe I don't get fully what you mean ?

anmonteiro replied:
Right, but it doesn't need to be exposed as a `module type`, creating yet another item in the public API. you can put the module type in a `ssl_intf.ml` file and reference it in the signature, for example. I think this talks about it https://www.craigfe.io/posts/the-intf-trick

anmonteiro commented on src/ssl.mli:
```diff
@@ -471,51 +445,96 @@ val set_ip : socket -> string -> unit
-val read : socket -> Bytes.t -> int -> int -> int
+(** {2 I/O on SSL sockets} *)
+
+(** The main communication functions below comes with
+    - a "Acquire" version that acquire the runtime while the operation takes place.
+    - a "NoAcquire" version that does not acquire the runtime while the operation takes place.
+
+    They come in two separate module with the same interface.
+
+    The acquire version are recommanded in one case:
+    - A program using several OCaml thread, if the socket is in blocking
+      more. In this case, as the global lock is released, the other OCaml
+      threads may run.
+
+    The No-acquire version is recommanded otherwise:
+    - i.e. the program is purely sequential (one thread)
+    - the socket is non blocking
+
+    Remark: whatever function you use with non-blocking socket, you must
+    capture all exception that requires you to retry later the same operation.
+*)
```
I rewrote the docstring and removed the verbiage around recommendations.

```suggestion
(** The main SSL communication functions are divided into the two following modules, which share the same interface:
    - `SslBlocking`, which does not release the OCaml runtime lock before calling the underlying SSL primitives.
    - `SslNonBlocking`, which releases the OCaml runtime lock while calling the underlying SSL primitives.

    The usual semantics apply to sockets in non-blocking mode, i.e. handling of `EWOULDBLOCK`,
    `EGAIN`, etc.
*)
```

anmonteiro reviewed (commented):
I added some comments and suggestions that I'd like to see addressed before we move forward.

Additionally, I'd love to see a different PR for the flush error change (3d7445d4551d67ff0ab285a9ced612b133ccc067) so it can be discussed separately. I did not review that part of the code.

craff commented:
> > Here are some tests with nginx+ssl (as reference) and simple_http+ssl on a 50MB file
> > number of connection	nginx-ssl	simple-ssl without PR	simple-ssl with PR
> > 5	50.30	36.56	59.33
> > 10	54.95	32.43	60.96
> > 15	54.63	33.81	61.60
> > 20	53.06	33.08	61.57
> > This PR brings almost a factor 2 and allows to be better that nginx!
> 
> What are the units here?

Request per seconds

craff commented:
> I added some comments and suggestions that I'd like to see addressed before we move forward.
> 
> Additionally, I'd love to see a different PR for the flush error change ([3d7445d](https://github.com/savonet/ocaml-ssl/commit/3d7445d4551d67ff0ab285a9ced612b133ccc067)) so it can be discussed separately. I did not review that part of the code.

- I reopen PR #104 for only fix_flush (and the warning/error for compilation of a test in 5.0) 
- I rebased this PR on #106 (meaning #104 should be merged first)
- I rebased PR #107 on #106.

You already reviewed #104 long ago and I just updated a comment.

craff commented:
Remark, there is a breaking change is this PR for the few "blocking" function that I replaced. We could redefine them from there equivalent in the SSLNoRelease module with a deprecated flag. What do you think ?

craff commented:
Ok, I understood, but if we do what you suggest the type of read and all other functions of the module will not be documented if SSL.mli.

Le 9 mai 2023 17:58:22 GMT-10:00, Antonio Nuno Monteiro ***@***.***> a écrit :
>@anmonteiro commented on this pull request.
>
>
>
>> +    They come in two separate module with the same interface.
>+
>+    The acquire version are recommanded in one case:
>+    - A program using several OCaml thread, if the socket is in blocking
>+      more. In this case, as the global lock is released, the other OCaml
>+      threads may run.
>+
>+    The No-acquire version is recommanded otherwise:
>+    - i.e. the program is purely sequential (one thread)
>+    - the socket is non blocking
>+
>+    Remark: whatever function you use with non-blocking socket, you must
>+    capture all exception that requires you to retry later the same operation.
>+*)
>+
>+module type SslCom = sig
>
>We're exposing a module type that has no use. I'd rather keep the API surface area as small as possible.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/savonet/ocaml-ssl/pull/106#discussion_r1189334363
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

anmonteiro commented:
> Ok, I understood, but if we do what you suggest the type of read and all other functions of the module will not be documented if SSL.mli. Le 9 mai 2023 17:58:22 GMT-10:00, Antonio Nuno Monteiro ***@***.***> a écrit :
> […](#)
> @anmonteiro commented on this pull request. > + They come in two separate module with the same interface. + + The acquire version are recommanded in one case: + - A program using several OCaml thread, if the socket is in blocking + more. In this case, as the global lock is released, the other OCaml + threads may run. + + The No-acquire version is recommanded otherwise: + - i.e. the program is purely sequential (one thread) + - the socket is non blocking + + Remark: whatever function you use with non-blocking socket, you must + capture all exception that requires you to retry later the same operation. +*) + +module type SslCom = sig We're exposing a module type that has no use. I'd rather keep the API surface area as small as possible. -- Reply to this email directly or view it on GitHub: [#106 (comment)](https://github.com/savonet/ocaml-ssl/pull/106#discussion_r1189334363) You are receiving this because you authored the thread. Message ID: ***@***.***>

Then I'd be OK repeating them in the `.mli`. I don't see a problem with that.

craff commented:
> > Ok, I understood, but if we do what you suggest the type of read and all other functions of the module will not be documented if SSL.mli. Le 9 mai 2023 17:58:22 GMT-10:00, Antonio Nuno Monteiro _**@**_.**_> a écrit :
> > […](#)
> > @anmonteiro commented on this pull request. > + They come in two separate module with the same interface. + + The acquire version are recommanded in one case: + - A program using several OCaml thread, if the socket is in blocking + more. In this case, as the global lock is released, the other OCaml + threads may run. + + The No-acquire version is recommanded otherwise: + - i.e. the program is purely sequential (one thread) + - the socket is non blocking + + Remark: whatever function you use with non-blocking socket, you must + capture all exception that requires you to retry later the same operation. +_) + +module type SslCom = sig We're exposing a module type that has no use. I'd rather keep the API surface area as small as possible. -- Reply to this email directly or view it on GitHub: [#106 (comment)](https://github.com/savonet/ocaml-ssl/pull/106#discussion_r1189334363) You are receiving this because you authored the thread. Message ID: _**@**_.***>
> 
> Then I'd be OK repeating them in the `.mli`. I don't see a problem with that.

All this made me think more ... I think we should even export SslBase and the functor in the mli. If you want to implement your own scheduler using algebraic effect, you just need to reimplement a structure of type SslBase from the SslNoRelease module. Could allow to provide a separate SslEio module/sub package easily. 

What do you think ?

craff commented:
I did a small commit to add back the two existing blocking function to this PR should not in principle introduce any breaking change. I marked them as deprecated. 

[0468519](https://github.com/savonet/ocaml-ssl/pull/106/commits/0468519e4ffdf7877f38df07b17ca257f2d3bd7c)

craff commented:
> > What do you think ?
> 
> I think that we should only add features once they're necessary. It was pretty straightforward to implement [`eio-ssl`](https://github.com/anmonteiro/eio-ssl) with the features of the library.
> 
> so my opinion is:
> 
>     * we should keep the APi surface area as small as possible
> 
>     * we should only expose what users will need, so 1. no functor 2. no module type 3.just the functions

Ok perfect.

craff commented:
> Then I'd be OK repeating them in the `.mli`. I don't see a problem with that.

SslCom is repaeted both in `Ssl.ml` and `Ssl.mli` So it is ok like that ?

Or you prefer that I remove SslCom and expand it twice as type of SslRelease and SslNoRelease ?

craff commented:
Thanks for accepting #104, I can now see my build workflow runing immediatly.

anmonteiro commented:
> SslCom is repaeted both in Ssl.ml and Ssl.mli So it is ok like that ?

No:
> so 1. no functor 2. no module type 3.just the functions

craff commented:
Remark: I think I could propose another PR with a new functor that take as input just an algebraic effect that that gives a structure of type SslCom which trigger the effect in case of WOULD_BLOCG/EAGAIN and provide a "retry" function of type (unit->unit) that always do what is needed. 

I could use that in Simple_httpd (where I can not use Eio because I use fdevent, edge triggering, ...  to have O(1) scheduling. I actually was considering getting my scheduler out of Simple_httpd because for application that have mostly IO base scheduling, it probably outperforms Eio (but is currently linux specific). 

I would do that only if you think such a PR as a chance, because I would like Simple_httpd to depend upon the current version of ocaml-ssl and not my own private fork.

anmonteiro commented:
Why would that need to live in `Ssl`? Can't it be implemented in user land?

craff commented:
> > SslCom is repaeted both in Ssl.ml and Ssl.mli So it is ok like that ?
> 
> No:
> 
> > so 1. no functor 2. no module type 3.just the functions

I hope [550098f](https://github.com/savonet/ocaml-ssl/pull/106/commits/550098fd997ac4dc1596a694051f3f1be934ddc7) is what you wished for

craff commented:
> Why would that need to live in `Ssl`? Can't it be implemented in user land?

The user would have to rewrite all functions in the SslCom module type (at least the one he needs) while we could provide a generic functor using a third time (but not exposing it) the functor SSlCome :  SslBase -> SslCom. If you think it could be interesting. Moreover, I think this is a very short piece of extra code.

I could submit a PR on which we can discuss ?

anmonteiro commented:
Feel free to submit the PR if you think it's a simple change, and we can discuss it.

craff commented:
> Feel free to submit the PR if you think it's a simple change, and we can discuss it.

I might do it but it will require an OCaml version supporting effect or a way to make it more abstract ... I will think about it.

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,40 @@ val input_int : socket -> int
+
+    This saves some buffer copy and can be up to 2x faster.
+*)
+module SslNoRelease : sig
```
This should be called `SslBlocking` or `Ssl_blocking`

craff replied:
> This should be called `SslBlocking` or `Ssl_blocking`

I disagree: SslNoRelease has at least two use cases with Blocking socket
- Mono threaded code (like a simple client)
- fork/mmap based parallelism (will probably disappear with 5.0 ?), which is mono threaded too.

Still I am not fully happy with the name.

craff replied:
> This should be called `SslBlocking` or `Ssl_blocking`

And more importantly, these are Blocking only if the socket is configured as blocking, so the name is misleading

a better name could be SslForNonBloackingSocket ?

anmonteiro replied:
I didn't read "blocking" as "socket in blocking mode", but rather "blocking" because it doesn't release the OCaml runtime (so it can't be "async". I was using "blocking" as the opposite of "async"). maybe "sync" is a better name then?

toots replied:
Blocking was the old terminology in ocaml c APIs and it was super confusing. The new terminology explicitly mentioned release/no release, i think it's generally much better.

craff replied:
It is the same SslSync is misleading, because it only describes the semantic you get with Blocking socket, while the module SslNoRelease can be used for Async semantics with non-blocking socket, which, I think, is its common use case. 

Naming by the semantics will not work because the semantics depends upon the socket type. This is why I ressorted to naming by what it does. Here is a list of proposition that could help us find a good name:

- SslNoRelease
- SslKeepLock
- SslForNonBlock (exclude sync mono threaded use case)
- SslDirect (also express that we avoid a buffer and that it is faster)

I like SslDirect: it is short and a fair description. You can not guess (wrongly) its meaning without looking at the documentation. Which is what we want for such an advanced feature ?

anmonteiro replied:
asked a friend for some more suggestions:
<img width="905" alt="image" src="https://github.com/savonet/ocaml-ssl/assets/661909/4605e0fe-c744-4648-ad91-0ed6ce22df49">

anmonteiro replied:
how about `Ssl.Runtime_locked` for the new set of functions? I think that's clear enough.

craff replied:
Yes, I like Ssl.RuntimeLocked (I prefer underscore for lowercase ident and capital for upper case ident in OCaml). Super :

craff replied:
[856a5c5](https://github.com/savonet/ocaml-ssl/pull/106/commits/856a5c57ce08cc1878acadc5f1409eb87d19eb19)
introduces the new name!

anmonteiro replied:
I realize it's your personal preference to use camel-casing, but there are no camel-cased names in this library, so we should probably preserve that property to minimize the surprise factor.

craff replied:
You are clearly right. 

[b5c6966](https://github.com/savonet/ocaml-ssl/pull/106/commits/b5c6966657735806ac6096fd349c294a3ce2dd76) solves this

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,40 @@ val input_int : socket -> int
 (** Write an integer on an SSL socket. *)
 val output_int : socket -> int -> unit
+
+(** In `SslNoRelease`, you will find equivalent of the previous function that
+    do not release the OCaml runtime lock before calling the underlying SSL
+    primitives. This should be preferer for non blocking sockets where the
+    usual semantics apply, i.e. handling of `EWOULDBLOCK`, `EGAIN`, etc.
+    In could also be used with blocking socket if you program has only one thread.
+
+    This saves some buffer copy and can be up to 2x faster.
```
```suggestion
(** [Ssl_blocking] is an equivalent, signature compatible, equivalent to the
    [Ssl] module, with one difference: the OCaml runtime lock isn't released
    before calling the underlying SSL primitives.
    
    It works well with non blocking sockets where the usual semantics apply,
    i.e. handling of `EWOULDBLOCK`, `EGAIN`, etc.
```

craff replied:
I will address this when we find a name we both agree.

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,40 @@ val input_int : socket -> int
+    This saves some buffer copy and can be up to 2x faster.
+*)
+module SslNoRelease : sig
+  val connect : socket -> unit
```
it shouldn't be too much work to copy the docstrings here too, and it'll provide a better experience for users.

craff replied:
> it shouldn't be too much work to copy the docstrings here too, and it'll provide a better experience for users.

Copying means the user should read both to see there is no diff, no copying means you are sure it is the same documentation ? It also keeps the size of mli minimal ?

anmonteiro replied:
Copying the docstring inline means the users will be able to see the docs for every function from their editor integration.

craff replied:
I forgot about editor integration! I don't use it. I guess exporting the module type would solve that? There is no [@@ same doc as ...] which could be nice. I think a bit more and probably will copy the docstring.

craff replied:
Fixed by [358177a](https://github.com/savonet/ocaml-ssl/pull/106/commits/358177a7952dbede5f6bec34abeab735cf4f8a36)

anmonteiro commented on src/ssl.ml:
```diff
@@ -265,91 +263,152 @@ external set_host : socket -> string -> unit = "ocaml_ssl_set1_host"
+
+(* The functor implementing communication functions from a structure of type
+   SslComBase *)
+module SslCom(C:SslBase) = struct
```
this should be called `Make` or `Ssl_make`. `SslCom` is a bit unclear.

craff replied:
> this should be called `Make` or `Ssl_make`. `SslCom` is a bit unclear.

Adressed by [28ad4b5](https://github.com/savonet/ocaml-ssl/pull/106/commits/28ad4b5b6aaaca38093c37404ab95fc4cfae4ca6). I chose SslMake ;-)

anmonteiro commented on src/ssl.ml:
```diff
@@ -265,91 +263,152 @@ external set_host : socket -> string -> unit = "ocaml_ssl_set1_host"
+module SslNoRelease = SslCom(SslNoReleaseBase)
+
+(* The releasing functions are imported as default *)
+include SslRelease
```
we can save a line here by applying the functor inline:

```ocaml
include Make_SSL (SslReleaseBase)
```

craff replied:
Addressed by [28ad4b5](https://github.com/savonet/ocaml-ssl/pull/106/commits/28ad4b5b6aaaca38093c37404ab95fc4cfae4ca6)

anmonteiro reviewed (commented):
This is starting to shape up really nicely. Thanks for your continued work.

craff commented:
> This is starting to shape up really nicely. Thanks for your continued work.

Thanks to you for the review!

craff commented:
I needed [856a5c5](https://github.com/savonet/ocaml-ssl/pull/106/commits/856a5c57ce08cc1878acadc5f1409eb87d19eb19) because of a recent merge in master ?

craff commented:
About the PR for deprecation of old protocol: I will redo it, it is a very small change and the history is too messy for that now with the rebase (git is really messy for merge/rebase compared to darcs ;-). You can wait to review that I do the new one please.

anmonteiro commented on src/ssl.ml:
```diff
@@ -265,91 +263,149 @@ external set_host : socket -> string -> unit = "ocaml_ssl_set1_host"
+end
+
+(* The functor implementing communication functions from a structure of type
+   SslComBase *)
```
```suggestion
   Ssl_base *)
```

anmonteiro commented on src/ssl.ml:
```diff
@@ -265,91 +263,149 @@ external set_host : socket -> string -> unit = "ocaml_ssl_set1_host"
+
+(* The functor implementing communication functions from a structure of type
+   SslComBase *)
+module Make(C:Ssl_base) = struct
```
```suggestion
module Make(Ssl_base: Ssl_base) = struct
```

anmonteiro commented on src/ssl.ml:
```diff
@@ -265,91 +263,149 @@ external set_host : socket -> string -> unit = "ocaml_ssl_set1_host"
+(* The functor implementing communication functions from a structure of type
+   SslComBase *)
+module Make(C:Ssl_base) = struct
+  include C
```
```suggestion
  include Ssl_base
```

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,83 @@ val input_int : socket -> int
+
+(** Deprecated functions for compatibility with older version *)
+val read_into_bigarray_blocking : socket -> bigarray -> int -> int -> int
+  [@@@ocaml.alert deprecated "Use [Runtime_lock.read_into_bigarray] instead"]
```
```suggestion
  [@@ocaml.alert deprecated "Use [Runtime_lock.read_into_bigarray] instead"]
```

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,83 @@ val input_int : socket -> int
+val read_into_bigarray_blocking : socket -> bigarray -> int -> int -> int
+  [@@@ocaml.alert deprecated "Use [Runtime_lock.read_into_bigarray] instead"]
+val write_bigarray_blocking : socket -> bigarray -> int -> int -> int
+  [@@@ocaml.alert deprecated "Use [Runtime_lock.write_bigarray] instead"]
```
```suggestion
  [@@ocaml.alert deprecated "Use [Runtime_lock.write_bigarray] instead"]
```

craff commented:
Many thanks for your work and latest commit.

anmonteiro commented on src/ssl.ml:
```diff
@@ -265,91 +263,149 @@ external set_host : socket -> string -> unit = "ocaml_ssl_set1_host"
+  val write_bigarray : socket -> bigarray -> int -> int -> int
+end
+
+(* We now provide the implementation base communication function that release
+   the ocaml runtime lock, allowed multi-thread with blocking IO *)
```
```suggestion
(* Provide the base implementation communication functions that release
   the OCaml runtime lock, allowing multiple systhreads to execute concurrently. *)
```

anmonteiro commented on src/ssl.ml:
```diff
@@ -265,91 +263,149 @@ external set_host : socket -> string -> unit = "ocaml_ssl_set1_host"
+  external ssl_shutdown : socket -> bool = "ocaml_ssl_shutdown"
+end
+
+(* Same as above, but not releasing the lock: for non-blocking IO or
+   mono-threaded program *)
```
```suggestion
(* Same as above, but doesn't release the lock. *)
```

anmonteiro commented on src/ssl_stubs.c:
```diff
@@ -1783,6 +1872,24 @@ CAMLprim value ocaml_ssl_shutdown(value socket)
+  switch (ret) {
+    case 0:
+    case 1:
+      /* close(SSL_get_fd(SSL_val(socket))); */
```
```suggestion
```

anmonteiro commented on src/ssl.mli:
```diff
@@ -475,19 +445,46 @@ val set_ip : socket -> string -> unit
 
+(** {2 I/O on SSL sockets} *)
+
+(** The main SSL communication functions that could block (if socket are in blocking
+    mode) do release the ocaml runtime lock, which in some case require an extra copy     of data. Alternatives are provided below *)
```
```suggestion
(** The main SSL communication functions that can block if sockets are in blocking
    mode. This set of functions releases the OCaml runtime lock, which can require
    extra copying of application data. The module [Runtime_lock] provided below 
    lifts this limitation by never releasing the OCaml runtime lock. *)
```

anmonteiro commented on src/ssl.mli:
```diff
@@ -475,19 +445,46 @@ val set_ip : socket -> string -> unit
+(** Send close notify to the peer. This is SSL_shutdown(3).
+ *  returns [true] if shutdown is finished, [false] in case [close_notify]
+ *  needs to be called a second time. *)
+val ssl_shutdown : socket -> bool
```
This wasn't previously exposed.

```suggestion
```

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,83 @@ val input_int : socket -> int
+  (** Send close notify to the peer. This is SSL_shutdown(3).
+   *  returns [true] if shutdown is finished, [false] in case [close_notify]
+   *  needs to be called a second time. *)
+  val ssl_shutdown : socket -> bool
```
This function wasn't previously exposed.
```suggestion
```

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,83 @@ val input_int : socket -> int
+    [Ssl] module, with one difference: the OCaml runtime lock isn't released
+    before calling the underlying SSL primitives.
+
+    It works well with non blocking sockets where the usual semantics apply,
+    i.e. handling of `EWOULDBLOCK`, `EGAIN`, etc.
```
```suggestion
    It works well with non blocking sockets where the usual semantics apply,
    i.e. handling of `EWOULDBLOCK`, `EGAIN`, etc. Additionally, the functions
    in this module don't perform a copy of application data buffers.
```

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,83 @@ val input_int : socket -> int
 val output_int : socket -> int -> unit
+
+(** [Runtime_lock] is an equivalent, signature compatible, equivalent to the
+    [Ssl] module, with one difference: the OCaml runtime lock isn't released
+    before calling the underlying SSL primitives.
```
```suggestion
    [Ssl] module, with one difference: the OCaml runtime lock isn't released
    before calling the underlying SSL primitives. Multiple systhreads cannot,
    therefore, run concurrently.
```

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,83 @@ val input_int : socket -> int
+  val output_int : socket -> int -> unit
+end
+
+(** Deprecated functions for compatibility with older version *)
```
```suggestion
(** This function is deprecated. Use [Runtime_lock.read_into_bigarray] instead. *)
```

anmonteiro commented on src/ssl.mli:
```diff
@@ -523,3 +514,83 @@ val input_int : socket -> int
+(** Deprecated functions for compatibility with older version *)
+val read_into_bigarray_blocking : socket -> bigarray -> int -> int -> int
+  [@@ocaml.alert deprecated "Use [Runtime_lock.read_into_bigarray] instead"]
+val write_bigarray_blocking : socket -> bigarray -> int -> int -> int
```
```suggestion
(** This function is deprecated. Use [Runtime_lock.write_bigarray] instead. *)
val write_bigarray_blocking : socket -> bigarray -> int -> int -> int
```

anmonteiro commented on src/ssl_stubs.c:
```diff
@@ -1377,15 +1377,12 @@ CAMLprim value ocaml_ssl_embed_socket(value socket_, value context)
 
   if (socket < 0)
     caml_raise_constant(*caml_named_value("ssl_exn_invalid_socket"));
-  caml_release_runtime_system();
```
I hadn't noticed this before. Do we need a new embed socket function? We should preserve the old behavior for `ocaml_ssl_embed_socket`.

craff replied:
I first duplicated this function, but looking at the documentation, I found that SSL_new and SSL_set_fd can not block and do not do any communication, so one function not acquiring the runtime is enough (or one acquiring it) 

But you are right, a lot of function like ocaml_ssl_get_file_descr, should probably not release the runtime lock neither. But maybe you want this in another PR and we currently keep the previous one.

I just reverted (commit [606c1e5](https://github.com/savonet/ocaml-ssl/pull/106/commits/606c1e58286c6c94ce851e7206b725b710ddec2a)) and will submit an issue now.

craff replied:
I just openned a new issue #113, if you could tell me what you think, I might propose a PR very quickly (or not at all if you don't like any of the two ideas ;-) 

More generally, there is a lot of unanswered issues on ocaml-ssl, this is not a very good for the image of the project. May be some of them desserve to be closed, other to have at least a short answer. I could actually answer some of them maybe ?

anmonteiro replied:
Good point. Some issue gardening is probably in order. Feel free to answer them, of course.

I like your idea for the latest one, I'll answer there.

craff replied:
I will first fix issue #103 (and even output_char), I think this is a serious bug (but probably no one is using them ?). I might look at all the other occurrences of ``ignore`` while I am at it.

Many thanks again for your review.

craff replied:
I commented a few issues (only one has no answer now) and probably #31 and #42 can be closed.

craff commented:
I spotted a odoc problem in ssl.mli fixed by [606c1e5](https://github.com/savonet/ocaml-ssl/pull/106/commits/606c1e58286c6c94ce851e7206b725b710ddec2a)

anmonteiro commented:
Thank you!

craff commented:
Antonio Nuno Monteiro writes:

> Thank you!

Thanks to you too!

> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.


--
Christophe Raffalli
tél: +689 87 23 11 48
web: http://raffalli.eu

Ce mail est signé avec pgp (Pièce jointe signature.asc, clef sur https://raffalli.eu/pgp)
This mail is signed pgp (Attachment signature.asc, clef on https://raffalli.eu/pgp)

toots commented:
Awesome y'all, that looks like a great collaboration! Looking forward to testing it!

craff commented:
I did more work and PRs ... and found/fixed more bugs (2 in my recent PRs, 2 in
previous code, 1 in previous code which has already issue #103).

When you have time, you should first review #116 (one bug fix) and #118
(4 bugs fixes).

I may split #118 in four PRs if you prefer.

Cheers,
Christophe

Antonio Nuno Monteiro writes:

> Thank you!
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.


--
Christophe Raffalli
tél: +689 87 23 11 48
web: http://raffalli.eu

Ce mail est signé avec pgp (Pièce jointe signature.asc, clef sur https://raffalli.eu/pgp)
This mail is signed pgp (Attachment signature.asc, clef on https://raffalli.eu/pgp)
